sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/RedundantMethodsInDataClassesCheck.kt

```diff
@@ -0,0 +1,186 @@
+    name = HASHCODE_METHOD_NAME
+    withNoArguments()
+    returnType = INT_TYPE
+}
+
+private val OBJECT_ARRAY_HASH_MATCHER = listOf(
+        FunMatcher(qualifier = "java.util.Objects", name = "hash"),
+        FunMatcher(qualifier = "java.util.Arrays", name = "hashCode")
+)
+
+@Rule(key = "S6207")
+class RedundantMethodsInDataClassesCheck : AbstractCheck() {
+
+    private val issueMessage = "Remove this redundant method which is the same as a default one."
+
+    override fun visitClass(klass: KtClass, context: KotlinFileContext) {
+        if (!klass.isData()) return
+
+        // if class is data class it always contains constructor
+        val klassParameters = klass.primaryConstructor!!.valueParameters
+        var equalsMethod: KtNamedFunction? = null
+        var hashCodeMethod: KtNamedFunction? = null
+
+        klass.body?.functions?.forEach {
+            when {
+                EQUALS_MATCHER.matches(it, context.bindingContext) -> equalsMethod = it
+                HASHCODE_MATCHER.matches(it, context.bindingContext) -> hashCodeMethod = it
+            }
+        }
+
+        equalsMethod?.let {
+            if (it.equalsHasDefaultImpl(klassParameters, it.valueParameters, klass.name)) {
+                context.reportIssue(it.nameIdentifier!!, issueMessage)
+            }
+        }
+
+        hashCodeMethod?.let {
+            if (it.hashCodeHasDefaultImpl(klassParameters, context.bindingContext)) {
+                context.reportIssue(it.nameIdentifier!!, issueMessage)
+            }
+        }
+    }
+
+    private fun KtNamedFunction.hashCodeHasDefaultImpl(klassParameters: List<KtParameter>, bindingContext: BindingContext): Boolean {
+        return if (hasBlockBody()) {
+            val returnExpressions = collectDescendantsOfType<KtReturnExpression>()
+            if (returnExpressions.size > 1) return false
+            checkHashExpression(returnExpressions[0], bindingContext, klassParameters)
+        } else {
+            checkHashExpression(this, bindingContext, klassParameters)
+        }
+    }
+
+    private fun checkHashExpression(expression: KtExpression, bindingContext: BindingContext, klassParameters: List<KtParameter>): Boolean {
+        val callExpressions = expression.collectDescendantsOfType<KtCallExpression> {
+            OBJECT_ARRAY_HASH_MATCHER.any { e ->
+                e.matches(it, bindingContext)
+            }
+        }
+        if (callExpressions.size > 1 || callExpressions.isEmpty()) return false
+
+        if ((callExpressions[0].calleeExpression as KtNameReferenceExpression).getReferencedName() == "hash") {
+            if (callExpressions[0].valueArguments.size != klassParameters.size) return false
+            callExpressions[0].valueArguments.forEach {
+                findParameter(it.getArgumentExpression(), klassParameters) ?: return false
+            }
+        } else {
+            val arguments = callExpressions[0].valueArguments[0].collectDescendantsOfType<KtValueArgumentList>()
```

Comment:
I think it would be safer to actually get the `arrayOf(...)` by using `callExpressions[0].valueArguments[0].getArgumentExpression()` and check that it's an `arrayOf(...)` with proper arguments. This logic might give some FPs if accidentally a different function was used